architecture.tex

```diff
@@ -53,58 +55,190 @@
+integers with \texttt{Integer}, like \texttt{x + Integer(1)/2}, or using
+\texttt{x + Rational(1, 2)}. SymPy provides a function \texttt{S} which can be
+used to convert objects to SymPy types with minimal typing, such as \texttt{x
+  + S(1)/2}. This gotcha is a small downside to using Python
+directly instead of a custom DSL, and we consider it to be worth it for the
+advantages listed above.
 
 %%
 %% Assumptions
 \subsection{Assumptions}
 
+An important feature of the SymPy core is the assumptions system. The
+assumptions system allows users to specify that symbols have certain common
+mathematical properties, such as being positive, imaginary, or integer. SymPy
+is careful to never perform simplifications on an expression unless the
+assumptions allow them. For instance, the identity $\sqrt{x^2} = x$ holds if
+$x$ is nonnegative ($x>=0$). If $x$ is real, the identity $\sqrt{x^2}=|x|$
+holds. However, for general complex $x$, no such identity holds.
+
+By default, SymPy performs all calculations assuming that variables are
+complex valued. This assumption makes it easier to treat mathematical problems
+in full generality.
+
+\begin{verbatim}
+>>> x = Symbol('x')
+>>> sqrt(x**2)
+sqrt(x**2)
+\end{verbatim}
+
+By assuming symbols are complex by default, SymPy avoids performing
+mathematically invalid operations. However, in many cases users will wish to
+simplify expressions containing terms like $\sqrt{x^2}$.
+
+Assumptions are set on \texttt{Symbol} objects when they are created. For
+instance \texttt{Symbol('x', positive=True)} will create a symbol named
+\texttt{x} that is assumed to be positive.
+
+\begin{verbatim}
+>>> x = Symbol('x', positive=True)
+>>> sqrt(x**2)
+x
+\end{verbatim}
+
+Some common assumptions that SymPy allows are \texttt{positive},
+\texttt{negative}, \texttt{real}, \texttt{nonpositive}, \texttt{nonnegative},
+\texttt{real}, \texttt{integer}, and \texttt{commutative}~\footnote{If $A$ and
+  $B$ are Symbols created with \texttt{commutative=False} then SymPy will keep
+  $AB$ and $BA$ distinct.}. Assumptions on any object can be checked with the
+\verb|is_|\texttt{\textit{assumption}} attributes, like \verb|x.is_positive|.
+
+Assumptions are needed only to restrict a domain so that certain
+simplifications can be performed. It is not required to make the domain match
+the input of a function. For instance, one can create the object
+$\sum_{n=0}^m f(n)$ as \texttt{Sum(f(n), (n, 0, m))} without setting
+\texttt{integer=True} when creating the Symbol object \texttt{n}.
+
+The assumptions system additionally has deductive capabilities. The
+assumptions use a three-valued logic using the Python builtin objects
+\texttt{True}, \texttt{False}, and \texttt{None}. \texttt{None} represents the
+``unknown'' case, meaning either the given assumption can be either true or
+false under the given information, for instance, \verb|Symbol('x', real=True).is_real|
+will give \texttt{None} because a real symbol might be
```

Comment:
I was unsure if it would be desirable to rewrite the "examples" in this section as actual REPL examples, or to leave them in the text. If we rewrite them as REPL examples we'll definitely need to point this out.